cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,470 @@
+import shutil
+import warnings
+
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from .. import DirectConstraint
+from ..expressions.core import Expression, Comparison, Operator, BoolVal
+from ..expressions.globalconstraints import GlobalConstraint
+from ..expressions.globalfunctions import GlobalFunction
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl
+from ..expressions.utils import is_num, is_any_list, eval_comparison, argval, argvals, get_bounds
+from ..transformations.get_variables import get_variables
+from ..transformations.normalize import toplevel_list
+from ..transformations.decompose_global import decompose_in_tree
+
+"""
+    Interface to CP Optimizers API
+
+    CP Optimizer, also a feature of IBM ILOG Optimization Studio, is a software library of constraint programming tools 
+    supporting constraint propagation, domain reduction, and highly optimized solution search.
+
+    Documentation of the solver's own Python API: (all modeling functions)
+    https://ibmdecisionoptimization.github.io/docplex-doc/cp/docplex.cp.modeler.py.html#module-docplex.cp.modeler
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_cpo
+"""
+
+class CPM_cpo(SolverInterface):
+    """
+    Interface to CP Optimizers API
+
+    Requires that the 'docplex' python package is installed:
+    $ pip install docplex
+
+    docplex documentation:
+    https://ibmdecisionoptimization.github.io/docplex-doc/
+
+    You will also need to install CPLEX Optimization Studio from IBM's website.
+    There is a free community version available.
+    https://www.ibm.com/products/ilog-cplex-optimization-studio
+
+    See detailed installation instructions at:
+    https://www.ibm.com/docs/en/icos/22.1.2?topic=2212-installing-cplex-optimization-studio
+
+    Creates the following attributes (see parent constructor for more):
+    - cpo_model: object, CP Optimizers model object
+    """
+
+    _docp = None  # Static attribute to hold the docplex.cp module
+
+    @classmethod
+    def get_docp(cls):
+        if cls._docp is None:
+            import docplex.cp as docp  # Import only once
```

Comment:
very elegant... perhaps something we should do in the other solvers too? To discuss...

Reply:
I think so, not all solvers need it but gurobi does